### ISSUE TITLE
Native building (hxcpp, HL)

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#4912]
With the HL target starting to take shape we have to think about how to deal with the native compilation. I currently see two main options:
1. Use a custom build system like hxcpp.
2. Use a project/makefile generator such as [koremake](https://github.com/ktxsoftware/koremake).

Both approaches have advantages and disadvantages and while I have a clear favorite, I would like to hear some opinions on the matter first.

@hughsando @ncannasse @RobDangerous @jgranick @underscorediscovery

**Comment by Gama11** [comment]
I assume you meant to link to https://github.com/KTXSoftware/koremake there?

**Comment by Simn** [comment]
Yes, thanks, fixed!

**Comment by hughsando** [comment]
The hxcpp makefile (build.xml) is very easy to print out from haxe, and
depends only on neko.
But tradition would dictate that you make your own.

**Comment by ncannasse** [comment]
> But tradition would dictate that you make your own.

:-)

I would actually be interested to see if we can share a common set of tools / helpers to handle/help-with native build systems. The current fragmentation makes it hard to correctly support all workflows in IDE for instance.

**Comment by nadako** [comment]
IMO port koremake to haxe and run it in interp, so no dependencies at all. After that, implement build.xml output for it, so it can output it just like a VC project or a makefile.

**Comment by RobDangerous** [comment]
I can port koremake to Haxe/neko for that, not a problem. Benefit is of course IDE support.

**Comment by Simn** [comment]
> IMO port koremake to haxe and run it in interp, so no dependencies at all. After that, implement build.xml output for it, so it can output it just like a VC project or a makefile.

That's actually a very interesting idea, just leave the choice to the user. I appreciate hxcpp's build system for the fact that it "just works", but at the same time I sometimes would like to have proper project files for debugging and general IDE support.

**Comment by andyli** [comment]
I would like to point out that other then the haxe outputs, we also need to do native compilation for the runtime stuffs, including building Neko and HL VMs, the "standard" core ndlls, and let users to build their own ndlls...

It would be great if koremake (or whatever) is able to replace our Makefiles, vc projects, and the [install.neko](https://github.com/HaxeFoundation/neko/blob/master/src/tools/install.neko) script...

**Comment by ncannasse** [comment]
The "meta generator" capable of producing both IDE files (hxproj, vcproj) and build files (build.xml, Makefile, etc.) seems like quite powerful, but I wonder if we can make it actually work. What kind/amount of input data would be required? In which format?

**Comment by RobDangerous** [comment]
koremake can already create IDE files and makefiles. Input files look like https://github.com/KTXSoftware/Kore/blob/master/korefile.js - could be changed to use a plain config file instead of a scripting language but using a scripting language inside the build system is very useful/powerful.

**Comment by delahee** [comment]
1- I do hope we wond roll out another build chain but hey that's life.
2- wow. korefiles looks very coooool to me.
3- I still like build.xml even if it sometime is a bit obscure.




2016-03-08 13:37 GMT+01:00 Robert Konrad <notifications@github.com>:

> koremake can already create IDE files and makefiles. Input files look like
> https://github.com/KTXSoftware/Kore/blob/master/korefile.js - could be
> changed to use a plain config file instead of a scripting language but
> using a scripting language inside the build system is very useful/powerful.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/4912#issuecomment-193767290>
> .
>



-- 
David Elahee

**Comment by Simn** [comment]
If the compiler generates such a file we have to keep configurability in mind. I picture two files being generated: One containing the per-compilation data (files, configuration passed via -D or whatever) which may be overwritten on each compilation and one which "extends" that file and allows manual configuration.

**Comment by Justinfront** [comment]
Manual configuration is surely the https://github.com/KTXSoftware/Empty/blob/master/khafile.js file?
But I do feel with Kha that node dependancy is a bit of a pain, I like the hxml approach, node is fine for Kha but I think having haxe build systems that rely on installing node would not be a good, it's like introducing grunt, ant, soon your spending more time installing stuff that can go wrong than actually using code, although maybe that's not what your thinking?  For instance there is a move to using Android Studio among tools for the second stage of compile when targettting android, but for out of the box openfl (when it works) and installs everything for you from simple terminal commands and then compiles, well that's super nice with NME users installing openfl just to setup anrdoid! No need to learn how xcode or android studio work just to test some code on a phone for the first time, I mean you want to know the tech works well first.  Haxe helps reduce real learning curves and removes platform specific tool chains, the enterprise companies can normally use all the platform specific stuff if need be, but the indie developer - the haxe early adoptors want simple tool chains.  So I would not be infavour of Node grunt or similar dependancy for Haxe HL and C++.

**Comment by RobDangerous** [comment]
@Simn Things like one config file loading another config file to extend it is already supported as well.
@Justinfront I think we already decided not to use node at the very start of the discussion.

**Comment by hughsando** [comment]
It would be reasonably easy to add xcode, vcproj and to the extent that it
is native, android studio output to hxcpp - it already has all the flags,
dependencies and file stored in memory.
Even if it were experimental to start with, you would be compiling using
hxcpp from day one, and the proj files would work with existing xml files
(eg, linc files).  There has been nothing stopping anyone who wants this
badly enough from submitting a PR anytime in the last 5 years, since it is
just haxe code.  It is surely easier to do this than write a whole new
system.  But there is the tradition of course.

Neko vs haxe --run: hxcpp is a haxelib so neko is no extra dependency if
you are willing to use a neko based haxelib, so it is a question of things
that can be easily measured.  Threading and startup time being the most
obvious measures.

**Comment by underscorediscovery** [comment]
Please don't make a new one. 
Please don't move to one that is littered with things it shouldn't be.

Please don't _require_ generated project files. This removes control from the user, and you have to replicate every single feature of every single project file in order to not stomp on access to critical platforms features. Generating the project file destructively on build is a nightmare in practice, I would not recommend it. Having the generation _as an option_ is convenient, there are many battle tested and minimal existing solutions (genie, premake, countless others) too look at since project generation is not a new desire. However, removing the simple, minimal cli builds as the baseline is a bad idea (not that that was suggested, just that it's critical to Haxe native targets being actually useful in flexible requirements).

Starting with the hxcpp build tool is the only reasonable option immediate term imo, it's got years of practical and pragmatic solutions to all the existing use cases, and already answers the needs perfectly reasonably. 

Please don't rush into making choices that are divisive and aren't carefully considered, the build process is pretty important to people (obviously) and it directly affects our ability to deliver on our own needs and promises.

**Comment by RobDangerous** [comment]
Not so sure about it being easy. I think more than 50% of the time I spent on koremake overall I spent on getting the Visual Studio and Xcode files right. Something like that should be easy and it is for most IDEs but not for those. I used premake years ago. Switched to making my own because Visual Studio and Xcode support in that one was troublesome and very slow.

**Comment by underscorediscovery** [comment]
Yes premake was forked and continued by genie https://github.com/bkaradzic/GENie

**Comment by Simn** [comment]
The idea here is not to invent something new which is why I mentioned two *existing* solutions initially. Neither of them are perfect at the moment which is why we are discussing how to approach and possibly unify this.

And yes, nobody suggested getting rid of hxcpp's build tool in favor of something else. The point is that we'll have to do something for HL anyway, so we might as well discuss a good strategy. Saying to not make anything "new" is not helpful because we couldn't blindly drop in the hxcpp build tool either without at least some effort.

**Comment by RobDangerous** [comment]
Just please don't use a slow lua tool :)

**Comment by underscorediscovery** [comment]
There is a habit of HF lately pinging people in threads asking for opinions, and then responding as if that person was yelling through the window, everyone can all relax a bit :)

I think a unified native build config/stage is useful, in my understanding this is the whole purpose of hxcpp. Everything that I know of is using and leveraging the power of the hxcpp build tool already - which is the unified native building tool I prefer because it's part of the official native backend. Are there projects that aren't using hxcpp?

**Comment by RobDangerous** [comment]
There is at least one project which uses hxcpp but does not use the hxcpp build tool.

**Comment by underscorediscovery** [comment]
More specifically on topic:
In my view, starting at the data is most useful. 

The first question should be what data the compiler emits, rather than how it is consumed or who is consuming it. For hxcpp this is currently the `Build.xml` + `Options.txt` file that the haxe emits. 

The format should just be plain data that is easy to parse and consume and remain agnostic and unopinionated - it's just data. This allows the compiler to make informed decisions and remain easily maintained over the long term.

Any easily parseable format would work like json, xml, text files - it wouldn't matter since it's a transport format. What happens above that format would come after it existing, working in order of pragmatic needs.

**Comment by Simn** [comment]
The data format should always be a consideration, but I don't think it's top priority here. We only want to have one build tool, so as long as that tool knows how to handle the data it's good enough.

Maybe we should make it extra hard to parse so nobody gets the idea to write a custom build tool!

**Comment by underscorediscovery** [comment]
Is the idea that this (native) build tool would somehow replace all build workflows for all users outside of the )native target -> native output) stage?

**Comment by back2dos** [comment]
Sven is right that the data needs to be well defined. 

> Smart data structures and dumb code works a lot better than the other way around.
> _-The Cathedral and the Bazaar_

Or as Torvalds put it:

> …git actually has a simple design, with stable and reasonably well-documented data structures. In fact, I'm a huge proponent of designing your code around the data, rather than the other way around, and I think it's one of the reasons git has been fairly successful […] I will, in fact, claim that the difference between a bad programmer and a good one is whether he considers his code or his data structures more important. Bad programmers worry about the code. Good programmers worry about data structures and their relationships.

Building on that, I would go further and not marshall it in the first place. One would need backend-specific callbacks, that are similar to Context.onAfterGenerate(), only that they provide the data to the callee.

So let's consider something like:

```haxe
typedef FileList = {
  compilerflag: String,
  files:Array<{
    name:String,
    dependencies:Array<String>,
  }>
}

typedef CppBuild = {
  flags:Map<String, String>,
  files:Map<String, FileString>, //maybe this should be a typed object instead of a map?
  options:Map<String, String>, //or should this be merged with flags? or a typed object?
}

static function registerCppBuilder(builder:CppBuild->Void):Void;
```

The type `CppBuild->Void` is a universal interface that everyone can rely on. Writing the code to generate to write `Build.xml` and `Options.txt` and then invoke the build tool is trivial. Or dumping a JSON file as Sven suggested. Or generating an XCode project directly. Or whatever.

There are some details to deal with, such as what to do if `registerCppBuilder` is called on the wrong backend, how to resolve conflicting registrations, defining the default behavior (be it by just building the way it happens now).

But, as already suggested, defining the data is the crucial step.

**Comment by ncannasse** [comment]
Sorry if I put the thread in the wrong path by mentioning data, before discussing technical details and data format I think it's more useful discussing for now what are the current problems we're trying to solve.

From my own humble point of view, there are several build systems in the Haxe ecosystem for now that are doing pretty much the same thing, I think it would be nice if we could discuss how to unify that.  I would like to study hxcpp build xml to learn more about it and see if it's suitable but it's not very well documented, making it hard to see the amount of things it can actually do. I like the idea of being able to generate various formats for 3rd party tools, but I also want to be able to compile without having to generate first intermediate files. For instance I like how hxcpp can both build using installed software, and it would be nice if it could also generate VisualStudio project files or Makefile for easier debugging of generated code.

@underscorediscovery i'm not sure I understand your comment regarding "HF people" :) Having an open discussion and inviting people to voice their opinion, eventually reaching an agreement that will make users life easier is I think something that we should do more than less, given the amount of fragmentation that we have reached in just a few years. Please tell us if you think that's a problem.

**Comment by underscorediscovery** [comment]
@ncannasse in the interest of progress I will just skip over you adding to yet missing my point.

I have been working on documenting hxcpp build format here https://github.com/snowkit/hxcpp-guide , I still have quite a bit locally to tidy and upstream in the next weeks. Maybe that will help you in the mean time.

**Comment by hughsando** [comment]
Juraj, you have basically described the "FileGroup" (FileList) and
"Compiler" ("CppBuild") classes in hxcpp tool.
Whether you have an elaborate plugin system, or simple an "if" in a single
repository I guess is a question of taste.

As for the complexities of project generation, you should examine the use
cases.  For example, I can conceive of no possible reason why you would
actually want a "makefile".

Nme already generates xcode projects.  I agree it is a horrible format, but
Nme cheats.  It builds a single "external dependency" that takes the
configuration (debug/arm/x64) and runs haxe.
This gives you the workflow: change haxe code -> hit "compile" button ->
takes you too errors in the .hx file in the IDE.
For msvc, you might put a break point in some generated code, change a .hx
file and hit F5 then inspect the variables.
All this can be done by cheating.  Do you really need anything else? ie,
does visual studio need to know that A.cpp depends on B.h, or does it
simple just need to run haxe again because in fact both depend on A.hx
instead.

I guess if you really wanted a makefile, you could make a one-liner
(two-liner in make) that just runs haxe.

**Comment by Justinfront** [comment]
https://ninja-build.org/
you can target ninja from higher level build tools.
https://github.com/ninja-build/ninja/wiki/List-of-generators-producing-ninja-build-files
Find ./configure && ninja     from terminal ideal for building TM2 I don't need to even touch xcode

**Comment by hughsando** [comment]
I love the fact that ninja is used to build "parts of Android".  More tools used in a build environment is better, right?

**Comment by Justinfront** [comment]
just looking briefly at Luxe today, I need to use snowfall to install flow which does not work with latest haxe. If flow created a ninja file I could just get the file off someone on irc perhaps and hack it till flow was fixed.  Also every editor is calling a different tool to build haxe projects.  I really don't know if ninja would help but the hxcpp build tool is kind of very haxe specific and perhaps c++ specific something like ninja is probably not, so it's more likely to fit future language needs?

**Comment by Justinfront** [comment]
but I mainly brought it up because no nice to build textmate from terminal without needing to open xcode, and I presume for haxe it would also help us having to open visual studio or android studio just to build projects.

**Comment by RobDangerous** [comment]
Xcode, Visual Studio and Android Studio projects can all be build from the terminal without any extra tools. That is not the problem ninja solves - it is just about faster builds.

**Comment by Simn** [comment]
I'll close this as inconclusive, and by now not worth it anymore.

**Comment by ncannasse** [comment]
For the record I have plan to rely on hxcpp build tool to compile HL/C
sources, we need to find the time to work on it with @hughsando
